htmresearch/algorithms/sparsematrix_temporal_memory/basal_context_apical_context.py

```diff
@@ -0,0 +1,482 @@
+
+"""An implementation of TemporalMemory"""
+
+
+import operator
+
+import numpy as np
+
+from htmresearch.algorithms.synapse_learning import createSynapseLearningRules
+from htmresearch.support import numpy_helpers as np2
+from nupic.bindings.math import Random, SparseMatrixConnections
+
+
+EMPTY_UINT_ARRAY = np.array((), dtype="uint32")
+
+
+class TemporalMemory(object):
+  """
+  An alternate approach to apical dendrites. Every cell SDR is specific to both
+  the basal the apical input. Prediction requires both basal and apical support.
+  """
+
+  def __init__(self,
+               columnDimensions=(2048,),
+               basalInputDimensions=(),
+               apicalInputDimensions=(),
+               cellsPerColumn=32,
+               activationThreshold=13,
+               initialPermanence=0.21,
+               connectedPermanence=0.50,
+               minThreshold=10,
+               sampleSize=20,
+               permanenceIncrement=0.1,
+               permanenceDecrement=0.1,
+               predictedSegmentDecrement=0.0,
+               maxNewSynapseCount=None,
+               maxSynapsesPerSegment=-1,
+               maxSegmentsPerCell=None,
+               seed=42, **kwargs):
+
+    self.columnDimensions = columnDimensions
+    self.numColumns = self._numPoints(columnDimensions)
+    self.basalInputDimensions = basalInputDimensions
+    self.apicalInputDimensions = apicalInputDimensions
+
+    self.cellsPerColumn = cellsPerColumn
+    self.connectedPermanence = connectedPermanence
+    self.minThreshold = minThreshold
+
+    self.sampleSize = sampleSize
+    if maxNewSynapseCount is not None:
+      print "Parameter 'maxNewSynapseCount' is deprecated. Use 'sampleSize'."
+      self.sampleSize = maxNewSynapseCount
+
+    if maxSegmentsPerCell is not None:
+      print "Warning: ignoring parameter 'maxSegmentsPerCell'"
+
+    self.predictedSegmentDecrement = predictedSegmentDecrement
+    self.activationThreshold = activationThreshold
+
+    self.synapseLearning = createSynapseLearningRules(
+      sampleSize=sampleSize,
+      initialPermanence=initialPermanence,
+      permanenceIncrement=permanenceIncrement,
+      permanenceDecrement=permanenceDecrement,
+      maxSynapsesPerSegment=maxSynapsesPerSegment
+    )
+
+    self.basalConnections = SparseMatrixConnections(
+      self.numColumns*cellsPerColumn, self._numPoints(basalInputDimensions))
+    self.apicalConnections = SparseMatrixConnections(
+      self.numColumns*cellsPerColumn, self._numPoints(apicalInputDimensions))
+    self.rng = Random(seed)
+
+    self.activeCells = EMPTY_UINT_ARRAY
+    self.winnerCells = EMPTY_UINT_ARRAY
+    self.prevPredictedCells = EMPTY_UINT_ARRAY
+
+
+  def reset(self):
+    self.activeCells = EMPTY_UINT_ARRAY
+    self.winnerCells = EMPTY_UINT_ARRAY
+    self.prevPredictedCells = EMPTY_UINT_ARRAY
+
+
+  def compute(self,
+              activeColumns,
+              basalInput,
+              basalGrowthCandidates,
```

Comment:
Growth candidates are the "winner cells". Often we don't want to be able to grow to all active input bits. For example, in sequence memory we only grow to one cell per column. We'll probably want to do the same thing elsewhere, e.g. L2 only connecting to "winner" cells in L4, or L4 only connecting to winners in L6.